### PR TITLE
[CALCITE-6639] Optimization that pulls up predicates causes ASOF JOIN validation failure

### PR DESCRIPTION
**Pull request by mihaibudiu**

### CODE REVIEW

**Comment by mihaibudiu** [comment]
The macOS CI failed with a SIGBUS. I swear it's not the fault of this PR.

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java [hunk 1]
```diff
@@ -774,8 +774,6 @@ public RelOptPredicateList inferPredicates(
       case INNER:
       case LEFT:
       case ANTI:
-      case ASOF:
-      case LEFT_ASOF:
```
Comment (caicancai) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java:778]: So in these two situations, should an exception be thrown, or do we need to take other measures?

**Review by caicancai — APPROVED** [review]
LGTM, just one question

**Review by suibianwanwank — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java (hunk 1 above)
Comment (suibianwanwank) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java:778]: It will return EMPTY in the end, looks good.

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java:778]: > It will return EMPTY in the end, looks good.

thanks

**Pull request merged**